logstash-core/lib/logstash/java_pipeline.rb

```diff
@@ -244,18 +244,53 @@ def start_workers
 
       filter_queue_client.set_batch_dimensions(batch_size, batch_delay)
 
+      # First launch WorkerLoop initialization in separate threads which concurrently
+      # compiles and initializes the worker pipelines
+
+      worker_init_thread = []
+      worker_loops = Queue.new
+
+      pipeline_workers.times do |t|
+        thread = Thread.new(
+          lir_execution,
+          filter_queue_client,
+          @events_filtered,
+          @events_consumed,
+          @flushRequested,
+          @flushing,
+          @shutdownRequested,
+          @drain_queue) do |*args|
+
+          begin
+            worker_loop = org.logstash.execution.WorkerLoop.new(*args)
```

Comment:
Red flag: `WorkerLoop` has a `static final ThreadLocal<ThreadContext> THREAD_CONTEXT`, which may hold references to the threads we use for intialization, potentially keeping memory from being freed.

Come to think of it, we may be leaking already, since pipeline reloads would potentially be leaving references there too.

Reply:
Good catch! Will investigate.

Reply:
I am moving this conversation into main comments.

Reply:
moving this conversation into main comments.

Reply:
per discussion current fix should not affect this behaviour, will followup on another issue for further investigation.